src/main/java/dk/kb/labsapi/ImageExport.java

```diff
@@ -0,0 +1,391 @@
+            illustrationAsByteArray = baos.toByteArray();
+        } catch (IOException e) {
+            log.error("Failed to download illustration from " + url + " while reading bytes");
+        }
+        return illustrationAsByteArray;
+         */
+        try {
+            return IOUtils.toByteArray(url);
+        }  catch (IOException e) {
+            log.error("Failed to download illustration from " + url + " while reading bytes");
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Streams content of all URLs given in input list to an output stream that delivers a zip file of all images from the URLs.
+     * @param illustrationURLs List of URls to get content from.
+     * @param illustrationMetadata List of metadata for each image returned from the URL list. Used to construct filenames.
+     * @param output output stream which holds the outputted zip file.
+     */
+    private void illustrationURLSToStream(List<URL> illustrationURLs, List<IllustrationMetadata> illustrationMetadata, OutputStream output, Map<String, Object> metadataMap) throws IOException {
+        ZipOutputStream zos = new ZipOutputStream(output);
+        zos.setLevel(Deflater.NO_COMPRESSION);
+
+        // Add metadata file to zip
+        try {
+            addMetadataFileToZip(metadataMap, zos);
+        } catch (Exception e) {
+            String message = String.format(
+                    Locale.ROOT,
+                    "Exception adding metadata entry to ZIP with %s illustrationURLs and %s illustrationMetadatas",
+                    illustrationURLs == null ? "null" : illustrationURLs.size(),
+                    illustrationMetadata == null ? "null" : illustrationMetadata.size());
+            log.warn(message, e);
+            throw new IOException(message, e);
+        }
+
+        int count = 0;
+        try {
+            for (int i = 0; i < illustrationURLs.size() ; i++) {
+                byte[] illustration = downloadSingleIllustration(illustrationURLs.get(i));
+                String pageUuid = illustrationMetadata.get(i).getPageUUID();
+                addToZipStream(illustration, String.format(Locale.ROOT, "pageUUID_%s_illustration_%3d.jpeg", pageUuid, count), zos);
+                count += 1;
+            }
+            // Close the zip output stream
+            zos.close();
+
+        } catch (IOException e) {
+            log.error("Error adding illustration to ZIP stream.");
+            throw new IOException();
+        }
+    }
+
+    /**
+     * Add a metadata file to the created zip file before images are added to the file.
+     * @param metadataMap containing metadata for the given solr query.
+     * @param zos ZipOutputStream to write the metadata file to.
+     */
+    public void addMetadataFileToZip(Map<String, Object> metadataMap, ZipOutputStream zos) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectWriter writer = mapper.writer(new DefaultPrettyPrinter());
```

Comment:
The writer auto-closes the `ZipOutputStream` when is has finished writing (yes, it's a trap). See the two new unit tests at the bottom of `ImageExportTest.java` for a demonstration and a fix.

Reply:
Very nice rewriting of the unit test!